### DISCUSSION TITLE
可以移除 ViaCoroutine ?

### DISCUSSION THREAD

**Discussion by yanminhui** [alibaba/async_simple#162] [General]
### Search before asking

- [x] I searched the [issues](https://github.com/alibaba/async_simple/issues) and found no similar issues.

### What happened + What you expected to happen

每次调用 `RescheduleLazy` 为了恢复到调用前所在的上下文，会通过 `ViaAsyncAwaiter` 附带一个恢复的协程 `ViaCoroutine`。
这个恢复的协程的逻辑是否可以合并到 `FinalAwaiter`，避免需要多调用这个协程？

![async_simple](https://www.plantuml.com/plantuml/png/bLPDQ-Cm4BtxLuZsaXoS7s1CQ1heKkXXMNONbbACQazRKKSA-d0sFVItL_wCHfQTz6PByyPltioRv4PfAhNPbT6p-4-k_YYGP2iAk2BN3-prJV-3iXI7Q6-3MCxsb6jInso554Op3dU6vve9ZY8UwCTnyW8Tt35Eoys1CWqo0buGow9bSi_jaRr4R7v8mFY_Z6xKaUTz2kzkeBcLG3LCNliNGLeHjGLLDarOWHD5u57yU7rR6dXQXsCJagQREXkirz5gjOh26KzPN4Lhy3sln7iD7ZYv4HsLduo_H-WS9qmmsaM5-QHnt05dE4-Md8kCrfAsKp21deOts4nBojMBaBj5dMMP-1FG7f_LSpmROLK3CpG-ciYR4CQPfcMcZDhR8LfOjjrZHii3FQhmTsR4H-ddc-3fqklNS_a5Ad-3mfJGwTWqnUM3VEMzTnzmIdq5lHNSFXjQCnZdHnNwwSQrDGEPnkVAS45z4Gqq3aWmB-mbeTE67DoC_UO7MWz8q4PoKWKZ4z9UPnMrtrbKInnrCk2ShSbBlRxSiKFxhDLdZGvzNKDD6uR31RfBeu3WDefGFF8ARgsZOiXaOTSEYbw0yTHTXY3ZKVo6uKV2-xybZaHdLal7VQv9tl1Zoo4l8muJBkXAYiTVTVhotNKZGPaTr4hyFcVfKQW_rMdQq68VlSt29lUcq7s_H8R0CY2F1IWjnH7rp6vtyeb3dxp-Nj9wxnnjtHlRPVlNx0i0 "async_simple")

### Are you willing to submit a PR?

- [ ] Yes I am willing to submit a PR!

**Comment by ChuanqiXu9** [comment]
ViaCoroutine 的主要用途是 co_await 用户自定义的 awaiter 时可以切换回来。

**Comment by yanminhui** [comment]
> ViaCoroutine 的主要用途是 co_await 用户自定义的 awaiter 时可以切换回来。

```.cpp
SimpleExecutor e;

Lazy<void> bar() {}

Lazy<void> foo() {
    co_await bar.via(&e); // 这个不属于用户定义的，但每次都要附带 `ViaCoroutine` 协程。
}
```

**Comment by ChuanqiXu9** [comment]
> > ViaCoroutine 的主要用途是 co_await 用户自定义的 awaiter 时可以切换回来。
> 
> ```c++
> SimpleExecutor e;
> 
> Lazy<void> bar() {}
> 
> Lazy<void> foo() {
>     co_await bar.via(&e); // 这个不属于用户定义的，但每次都要附带 `ViaCoroutine` 协程。
> }
> ```

指的是非 Lazy 类型的其他 Awaiter

**Comment by yanminhui** [comment]
> 指的是非 Lazy 类型的其他 Awaiter

我的意思是像上面这种是否可以解决不需要附带 `ViaCoroutine`。

另，对于你说的非 Lazy 的 Awaiter，`detail::coAwait(ex, awaitable)` 是否可以根据是否存在 `ex` 来决定是否需要 `ViaCoroutine` ?

**Comment by ChuanqiXu9** [comment]
> > 指的是非 Lazy 类型的其他 Awaiter
> 
> 我的意思是像上面这种是否可以解决不需要附带 `ViaCoroutine`。

RescheduleLazy 按设计其实本来只有 ReschduleLazy.start() 这种用法的，`co_await RescheduleLazy;` 不在设计范围内，所以我觉得针对这种 case 进行优化并没必要。如果可以的话，可以加一些约束，让我们只能 DetachedCoroutine 中 `co_await RescheduleLazy`，我记得这样会好一些。我添加了一个相关的 issue：https://github.com/alibaba/async_simple/issues/163

> 
> 另，对于你说的非 Lazy 的 Awaiter，`detail::coAwait(ex, awaitable)` 是否可以根据是否存在 `ex` 来决定是否需要 `ViaCoroutine` ?

无法判断其他 awaiter 中是否有 ex。如果你说判断 ex 是否非 null 的话，那本来就会判断吧。

**Reply by yanminhui** [reply]
>> 另，对于你说的非 Lazy 的 Awaiter，detail::coAwait(ex, awaitable) 是否可以根据是否存在 ex 来决定是否需要 ViaCoroutine ?
>
> ~~无法判断其他 awaiter 中是否有 ex。~~ 如果你说判断 ex 是否非 null 的话，那本来就会判断吧。

```.cpp
// ViaCoroutine 添加日志
std::suspend_always initial_suspend() const noexcept {
  std::clog << "ViaCoroutine::initial_suspend()" << std::endl;
  return {};
}
FinalAwaiter final_suspend() noexcept {
  std::clog << "ViaCoroutine::final_suspend()" << std::endl;
  return FinalAwaiter(_ctx);
}

struct UserAwaiter {
  bool await_ready() const noexcept { return false; }
  auto await_suspend(std::coroutine_handle<> handle) { return handle; }
  void await_resume() const noexcept {}
};

int main() {
  auto coro = []() -> Lazy<void> {
    std::clog << "coro body" << std::endl;
    co_return co_await UserAwaiter{};
  };
  syncAwait(coro());
  return 0;
}
```

运行结果：

```.log
$ ./main 
coro body
ViaCoroutine::initial_suspend()
ViaCoroutine::final_suspend()
```

上述 lambda 协程是不带执行器的，结果显示仍然运行 `ViaCoroutine`。

**Reply by ChuanqiXu9** [reply]
是啊，这个行为是正确的，就是 ViaCoroutine 为其他没有实现 coAwait 的 awaiter 来进行调度的。我不太明白你的意思

**Reply by yanminhui** [reply]
`main` 里的协程没有拥有执行器，应当不需要 `ViaCoroutine`。
当它通过 `.via(&e)` 执行时才需要 `ViaCoroutine`。
目前的情况是哪种情况都会走 `ViaCoroutine`。

**Reply by ChuanqiXu9** [reply]
不对，我们的设计是没有 coAwait 时走 ViaCoroutine，有 coAwait 反而就不需要 ViaExecutor 了。有 coAwait 后会执行 coAwait 的返回的 awaiter。这里文档好像确实缺了，是得补一下。

**Reply by RainMark** [reply]
ViaCoroutine名字可能不太合适？ViaCoroutine用来避免其他自定义awaiter被调度到其他线程。其实和.via(&e)关系不大。。。

**Reply by ChuanqiXu9** [reply]
有可能，这里的 via 导致歧义了

**Reply by yanminhui** [reply]
> ViaCoroutine名字可能不太合适？ViaCoroutine用来避免其他自定义awaiter被调度到其他线程。其实和.via(&e)关系不大。。。

@RainMark 要是没有 `executor` 走 `ViaCoroutine` 也没用，它没有上下文信息，然后用户定义的 `awaiter` 切到其它线程去，也恢复不回来。

**Reply by ChuanqiXu9** [reply]
这里的 executor 指的是 context 的 executor 而不是 awaiter 的 executor

**Reply by yanminhui** [reply]
对，我们看下上面这个代码片段，虽然有一个 `executor`，但它是 `nullptr`，没有人给它贱值，假设上面那个 `UserAwaiter` 切到其它线程去 `resume`，经过 `ViaCoroutine` 也回不到原来所在的线程，这个 `ViaCoroutine` 看起来没有意义。

**Reply by ChuanqiXu9** [reply]
空 executor 的情况属于预期外行为，除了我们在测试中会使用 syncAwait 以外，实际运用时都要加上 via 指定调度器的。同时就算是这种情况，ViaCoroutine 也不会造成逾期外行为。目前看应该没有必要对这种 edge case 进行优化。

**Reply by RainMark** [reply]
其实不太建议用户去自定义awaiter，应该是用future awaiter去和系统现有的异步接口去对接

**Reply by RainMark** [reply]
ViaCoroutine更多是一种正确性保证。确保默认情况下协程经过了自定义awaiter就能正确回到之前的线程。毕竟业务逻辑上不一定都是允许同一个协程跨线程执行。如果业务逻辑确定可以lazy挂起恢复前后被任意调度，那么用户自定义executor可以不实现checkin即可满足。另外不传executor的确是预期外行为，因为lazy后处理逻辑全在回调线程里立即执行了，一方面会有lazy跨线程执行风险，另一方面大量后处理任务可能回调线程（例如io线程池组）
